sparkly/utils.py

```diff
@@ -59,6 +64,89 @@ def absolute_path(file_path, *rel_path):
     )
 
 
+def chunks(seq, n):
+    """Iterate over seq by chunks of n elements.
+
+    >>> list(chunks([1, 2, 3, 4, 5], 2))
+    [[1, 2], [3, 4], [5]]
+    >>> list(chunks([], 2))
+    []
+    >>> list(chunks([1], 2))
+    [[1]]
+    """
+    it = iter(seq)
+    while True:
+        chunk = list(islice(it, n))
+        if not chunk:
+            break
+        yield chunk
+
+
+def type_has(t, required_fields):
+    """Check whether a complex dataType has specific fields.
+
+    Args:
+        t (pyspark.sql.types.ArrayType, pyspark.sql.types.MapType or
+            pyspark.sql.types.StructType): type to check.
+        required_fields (same with t or dict[str: pyspark.sql.DataType]):
+            fields that need to be present in t. For convenience, a user
+            can define a dict in place of a pyspark.sql.types.StructType,
+            but other than that this argument must have the same type as
+            t.
+
+    Raises:
+        AssertionError, if t and required_fields cannot be compared
+           because they aren't instances of the same complex dataType.
+        KeyError, if a required field is not found in the struct.
```

Comment:
I like using default exceptions because they have a non-ambiguous, pre-defined meaning, and the user won't need to import any special exceptions to be able to catch them. Also I think these ones lend themselves naturally to the errors they signal in this case. Even spark structures like `Row` use them.

To your two points:
* nothing should go wrong in the function itself, but if it does it will look like it in the traceback. But this should happen extremely rarely.
* Feedback on the error is very clear imo. For key errors it's the same you get when a key doesn't exist in a dictionary (which is a similar operation) with us providing the full path to the missing field, and for type errors we explicitly state what we were expecting and what we saw.